lib/bridge/MemberListSyncer.js

```diff
@@ -162,20 +162,57 @@ MemberListSyncer.prototype._getSyncableRooms = function(server, ignoreCache) {
         let attempts = 0;
         while (true) { // eslint-disable-line no-constant-condition
             try {
-                // roomDict = { room_id: RoomInfo }
-                let roomDict = yield self.appServiceBot.getMemberLists();
-                return Object.keys(roomDict).map(function(roomId) {
-                    return roomDict[roomId];
-                }).filter(function(roomInfo) {
+                let roomInfoList = [];
+                let joinedRoomIds = yield self.appServiceBot.getJoinedRooms();
+                for (let i = 0; i < joinedRoomIds.length; i++) {
+                    let roomId = joinedRoomIds[i];
+                    try {
+                        let userMap = yield self.appServiceBot.getJoinedMembers(roomId);
+                        let roomInfo = {
+                            id: roomId,
+                            state: [], // JSON objects (Matrix events)
+                            realJoinedUsers: [], // user IDs
+                            remoteJoinedUsers: [], // user IDs
+                        };
+                        let userIds = Object.keys(userMap);
+                        for (let j = 0; j < userIds.length; j++) {
+                            let userId = userIds[j];
+                            // TODO: Make this function public, it's useful!
+                            if (self.appServiceBot._isRemoteUser(roomId, userId)) {
+                                roomInfo.remoteJoinedUsers.push(userId);
+                            }
+                            else {
+                                roomInfo.realJoinedUsers.push(userId);
+                            }
+
+                            roomInfo.state.push({
+                                room_id: roomId,
+                                state_key: userId,
+                                user_id: userId,
+                                content: {
+                                    membership: "join",
+                                    display_name: userMap[userId].display_name,
+                                    avatar_url: userMap[userId].avatar_url,
+                                }
+                            });
+                        }
+                        roomInfoList.push(roomInfo);
+                    }
+                    catch (err) {
+                        log.error(`Failed to getJoinedMembers in room ${roomId}: ${err}`);
+                        i--; // try again
```

Comment:
Retrying forever on a room that consistently produces errors should probably be avoided. Although, I can't think of a good reason why one room would fail, despite it technically being possible. WDYT?